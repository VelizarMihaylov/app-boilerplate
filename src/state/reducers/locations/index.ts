import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime'
import utc from 'dayjs/plugin/utc'
import customParseFormat from 'dayjs/plugin/customParseFormat'

export type Locations = Array<{
  location: string
  lastUpdated: string
  measurements: {
    parameter: string
    value: number
    unit: string
  }[]
}>

type Action =
  | {
      type: 'ADD_LOCATIONS'
      payload: Locations
    }
  | {
      type: 'REMOVE_LOCATION'
      payload: string
    }

const locations = (
  store: [] | Locations = [],
  action: Action
): [] | Locations => {
  switch (action.type) {
    case 'ADD_LOCATIONS':
      return action.payload.map(measurement => {
        dayjs.extend(utc)
        dayjs.extend(relativeTime)
        dayjs.extend(customParseFormat)
        const formattedDate = dayjs(measurement.lastUpdated)
          .utc()
          .fromNow()
        return {
          ...measurement,
          lastUpdated: formattedDate
        }
      })
    case 'REMOVE_LOCATION':
      return [...store.filter(({ location }) => location !== action.payload)]
    default:
      return store
  }
}

export default locations
