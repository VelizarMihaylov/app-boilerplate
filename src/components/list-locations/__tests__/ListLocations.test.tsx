import React from 'react'
import { ListLocations } from '..'
import { create, act } from 'react-test-renderer'

import locations from '../__mocks__'

import { useQuery } from '@apollo/react-hooks'
jest.mock('@apollo/react-hooks')

const mockedUseQuery = useQuery as jest.Mock<unknown>

import { useDispatch, useSelector } from 'react-redux'
jest.mock('react-redux')

const mockedUseDispatch = useDispatch as jest.Mock<unknown>
const mockedUseSelector = useSelector as jest.Mock<unknown>

describe('ListLocations', () => {
  afterEach(() => jest.resetAllMocks())

  it('should handle loading state', () => {
    mockedUseQuery.mockImplementation(() => ({ loading: true }))
    mockedUseDispatch.mockImplementation(() => jest.fn())
    mockedUseSelector.mockImplementation(() => ({
      city: 'LONDON',
      locations: []
    }))
    const ListLocationsRender = create(<ListLocations />).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })
  it('should handle error state', () => {
    mockedUseQuery.mockImplementation(() => ({ error: true }))
    mockedUseDispatch.mockImplementation(() => jest.fn())
    mockedUseSelector.mockImplementation(() => ({
      city: 'LONDON',
      locations: []
    }))
    const ListLocationsRender = create(<ListLocations />).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })
  it('should render locations when not in loading or error state and locations are set in the store', () => {
    const { latestMeasurements } = locations
    mockedUseQuery.mockImplementation(() => ({
      loading: false,
      error: false,
      data: []
    }))
    mockedUseDispatch.mockImplementation(() => jest.fn())
    mockedUseSelector.mockImplementation(() => ({
      city: 'LONDON',
      locations: latestMeasurements
    }))
    const ListLocationsRender = create(<ListLocations />).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })

  it('should dispatch ADD_LOCATIONS action with the locations received as data', () => {
    const { latestMeasurements } = locations
    mockedUseQuery.mockImplementation(() => ({
      loading: false,
      error: false,
      data: locations
    }))
    const dispatch = jest.fn()
    mockedUseDispatch.mockImplementation(() => dispatch)
    mockedUseSelector.mockImplementation(() => ({
      city: 'LONDON',
      locations: []
    }))
    const ListLocationsRender = create(<ListLocations />)

    act(() => ListLocationsRender.update(<ListLocations />))
    expect(dispatch).toHaveBeenCalledWith({
      type: 'ADD_LOCATIONS',
      payload: latestMeasurements
    })
  })
})
