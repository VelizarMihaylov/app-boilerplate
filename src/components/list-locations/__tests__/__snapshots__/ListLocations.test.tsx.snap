// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListLocations should handle error state 1`] = `
<h1>
  Oops something went wrong!
</h1>
`;

exports[`ListLocations should handle loading state 1`] = `
<div
  className="css-92heyk"
>
  <div
    className="css-oajtzp"
  />
</div>
`;

exports[`ListLocations should render locations when not in loading or error state and locations are set in the store 1`] = `
<div
  className="css-92heyk"
>
  <div
    className="css-m8ffne"
  >
    <svg
      className="css-17hessb"
      id="Capa_1"
      onClick={[Function]}
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="css-xbce0x"
    >
      <p
        className="css-b0bjy"
      >
        Updated 2019-12-28T08:00:00.000Z
      </p>
      <h3>
        Manchester Piccadilly
      </h3>
      <p>
        in LONDON, United Kingdom
      </p>
      <p
        className="css-66vicj"
      >
        Value: PM25: 10 µg/m³, SO2: 11 µg/m³, NO2: 22 µg/m³, O3: 17 µg/m³
      </p>
    </div>
  </div>
  <div
    className="css-m8ffne"
  >
    <svg
      className="css-17hessb"
      id="Capa_1"
      onClick={[Function]}
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="css-xbce0x"
    >
      <p
        className="css-b0bjy"
      >
        Updated 2019-12-28T08:00:00.000Z
      </p>
      <h3>
        Manchester Sharston
      </h3>
      <p>
        in LONDON, United Kingdom
      </p>
      <p
        className="css-66vicj"
      >
        Value: O3: 34 µg/m³, NO2: 10 µg/m³
      </p>
    </div>
  </div>
  <div
    className="css-m8ffne"
  >
    <svg
      className="css-17hessb"
      id="Capa_1"
      onClick={[Function]}
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="css-xbce0x"
    >
      <p
        className="css-b0bjy"
      >
        Updated 2019-12-28T08:00:00.000Z
      </p>
      <h3>
        Salford Eccles
      </h3>
      <p>
        in LONDON, United Kingdom
      </p>
      <p
        className="css-66vicj"
      >
        Value: NO2: 13 µg/m³, PM25: 13 µg/m³, PM10: 16 µg/m³
      </p>
    </div>
  </div>
</div>
`;
